#1453 Make sure that we always create non bound collection

filiphr opened the pull request:
When generating collection mapping methods make sure that the created variable is from the bound type

Fixes #1453

sjaakd commented on processor/src/main/resources/org/mapstruct/ap/internal/model/IterableMappingMethod.ftl:
```diff
@@ -121,7 +121,7 @@
             <@includeModel object=resultType.implementationType/>
         <#else>
-            <@includeModel object=resultType/>
+            <@includeModel object=resultType withoutBounds=true/>
```
I'm not sure this is a good idea.. We are generating the methods ourselves because of nesting right? Should we even generate a method with a wildcard return type? See [here](http://www.angelikalanger.com/GenericsFAQ/FAQSections/ProgrammingIdioms.html#Should%20I%20use%20wildcards%20in%20the%20return%20type%20of%20a%20method?) for some details.. Just google "avoid return type wildcard". Of course.. we probably need to do something like this if the user decides to define a mapping method returning a wildcard.

filiphr replied:
Don't have access to that. In any case we are delivering `Collection<? extends MyType>`. If you all at [`Collection#addAll(Collection)`](https://docs.oracle.com/javase/8/docs/api/java/util/Collection.html#addAll-java.util.Collection-) it accepts `Collection<? extends E>` that is what the setter does actually too internally

sjaakd replied:
Fixed the link.. Perhaps I'm wrong.. I'll try to make an example

sjaakd commented on processor/src/test/java/org/mapstruct/ap/test/bugs/_1453/Auction.java:
```diff
@@ -0,0 +1,43 @@
+        this.otherPayments = otherPayments;
+    }
+
+    public List<Payment> getPayments() {
```
What about Maps? Would we have a similar case there? Or even other parameterised types..?

sjaakd commented:
Hi Filip, 

Try this:
```java
public class NewClass {

    public static void main(String[] args) {

        List<MyBase> baseCollection = new ArrayList<>();
        MyBase base = baseCollection.get( 0 );
        add1( baseCollection );
        add2( baseCollection );
        List<MyExt1> ext1Collection = new ArrayList<>();
        MyExt1 ext1 = ext1Collection.get( 0 );
        add1( ext1Collection );
        add2( ext1Collection ); // error.. MyBase only has two supertypes: Object and MyBase itself.

    }

    public static void add1(List<? extends MyBase> list ) {        
    }
    
    public static void add2(List<? super MyBase> list ) {
    }
    
    public class  MyBase {
    }
    
    public static class MyExt1 extends MyBase {        
    }
    
    public static class MyExt2 extends MyBase {     
    }
}
```
You should be fine with the `public setPayment(Collection<? extends Payment> payments)`, adding `Collection<Payment>` according to the above example. The same goes for `public setPayment(Collection<? super Payment> payments)`. `Payment` itself belongs both, to the 'family' of objects that is both a super of `Payment` and is a sub of `Payment`

The [FAC](http://www.angelikalanger.com/GenericsFAQ/FAQSections/ProgrammingIdioms.html) on generics is my bible when it comes to generics :smile:. 

I personally think this stuff is tricky. So when using wild card types defined types (so not calling method argument like above, but handling it in the body of this method), this [SO](https://stackoverflow.com/questions/4343202/difference-between-super-t-and-extends-t-in-java) answer is also really helpful. The *PECS* acronym helps me to remember.

filiphr commented:
@sjaakd you are correct. However, you are mixing the types. We would never be in such situation. We are always generating a method that would map to the one defined in the setter. Which means that our extension would be `? extends MyBase` or `? super MyBase`. This would lead to us creating an inner collection of the bound so `MyBase`.

What do you think I need to change in this PR?

filiphr commented:
>However, my example was meant to demonstrate that our return type for forged methods which does not need to follow the exact `? extends MyBase` or `? super MyBase`

Now it makes a lot more sense to me 😄. You are completely right, maybe we need to create a method which would return just the type bound. That way we don't even need the stuff I did in this PR. I will try that later this week in his PR

sjaakd commented:
> Now it makes a lot more sense to me 😄. You are completely right, maybe we need to create a method which would return just the type bound. That way we don't even need the stuff I did in this PR. I will try that later this week in his PR

Exactly.. but only for the forged method.. For implemented methods your solution will still be valid..

filiphr commented:
@sjaakd I've updated the PR.  I scrubbed my previous work. It seems doing it only for forged methods is enough. We actually don't allow implementing a method whose return type has an extends bound (and those are the ones that make problems). I've added a fixture as well so we can see how it looks like.

filiphr commented:
>We can't forbid a user to define a mapping method with extends and super bounds in the return type.

Well we are actually doing that for extends. See [this](https://github.com/mapstruct/mapstruct/blob/master/processor/src/main/java/org/mapstruct/ap/internal/util/Message.java#L116).

And for super there is no problem as this is valid:

```java
List<? super AuctionDto> list = new ArrayList<AuctionDto>( auctions.size() );
for ( Auction auction : auctions ) {
    list.add( map( auction ) );
}
```

See the fixture in the PR as well.

filiphr commented:
It seems has been done as part of #527 from you 😉

sjaakd commented:
> It seems has been done as part of #527 from you 😉

Hmm. my memory is failing me here.. must be my age 😢 . I have no further comments :smile:

filiphr commented:
Thanks for the review @sjaakd
